de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java

```diff
@@ -1,529 +1,863 @@
-            return;
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Reacting before resource move - resource: " + oldFile
+                + ", old parent: " + oldParent + ", new Parent: " + newParent);
         }
 
-        path = makeAbsolutePathProjectRelative(path, project);
+        if (oldFile.isDirectory()) {
+            generateFolderMove(oldFile, oldParent, newParent, null);
 
-        SPath spath = new SPath(project, path);
-        User user = resourceManager.getSession().getLocalUser();
-
-        IActivity activity;
-        if (virtualFileEvent.getFile().isDirectory()) {
-            activity = new FolderDeletedActivity(user, spath);
         } else {
-            activity = new FileActivity(user, Type.REMOVED, Purpose.ACTIVITY,
-                spath, null, null, null);
+            generateFileMove(oldFile, oldParent, newParent, null, null);
         }
-
-        project.removeResource(path);
-        editorManager.removeAllEditorsForPath(spath);
-
-        resourceManager.internalFireActivity(activity);
     }
 
-    @Override
-    public void fileMoved(
+    /**
+     * Generates and dispatches matching creation and deletion activities
+     * replicating moving the given directory. Also dispatches activities for
+     * all contained resources belonging to the same module.
+     * <p></p>
+     * How the resources (including the given directory) are handled depends on
+     * whether the source and target directory is shared.
+     * <ul>
+     * <li>
+     * If only the source directory is shared, only deletion activities for the
+     * old resources are dispatched.
+     * </li>
+     * <li>
+     * If only the target directory is shared, only creation activities for the
+     * new resources are dispatched.
+     * </li>
+     * <li>
+     * If both directories are shared, the resources are moved as follows:
+     * <p>
+     * Creation activities for the new directories (located in the target
+     * directory) and move activities for the contained files are dispatched in
+     * increasing depth order. Then, deletion activities for the old directories
+     * (in the source directory) are dispatched in decreasing depth order. This
+     * ensures that any resource move is already handled correctly before one of
+     * its parent resources is deleted.
+     * </li>
+     * </ul>
+     *
+     * <p></p>
+     * Renaming directories is also handled as a move activity. This can be done
+     * with the optional parameter <code>newFolderName</code>.
+     *
+     * @param oldFile       the directory that is about to be moved
+     * @param oldParent     the old parent of the moved file
+     * @param newParent     the new parent of the moved file
+     * @param newFolderName the new name for the folder or null if the folder
+     *                      was not renamed
+     * @see #generateFileMove(VirtualFile, VirtualFile, VirtualFile, String, String)
+     */
+    private void generateFolderMove(
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=307&issues=AWcYB0JQkEeD8QRk19vS&open=AWcYB0JQkEeD8QRk19vS)

[ISSUE_KEY]: AWcYB0JQkEeD8QRk19vS